compiler/rustc_ast_lowering/src/lifetime_collector.rs

```diff
@@ -10,14 +12,44 @@ use rustc_span::symbol::{Ident, kw};
 use super::ResolverAstLoweringExt;
 
 struct LifetimeCollectVisitor<'ast> {
-    resolver: &'ast ResolverAstLowering,
+    resolver: &'ast mut ResolverAstLowering,
+    always_capture_in_scope: bool,
     current_binders: Vec<NodeId>,
     collected_lifetimes: FxIndexSet<Lifetime>,
 }
 
 impl<'ast> LifetimeCollectVisitor<'ast> {
-    fn new(resolver: &'ast ResolverAstLowering) -> Self {
-        Self { resolver, current_binders: Vec::new(), collected_lifetimes: FxIndexSet::default() }
+    fn new(resolver: &'ast mut ResolverAstLowering, always_capture_in_scope: bool) -> Self {
+        Self {
+            resolver,
+            always_capture_in_scope,
+            current_binders: Vec::new(),
+            collected_lifetimes: FxIndexSet::default(),
+        }
+    }
+
+    fn visit_opaque(&mut self, opaque_ty_node_id: NodeId, bounds: &'ast GenericBounds, span: Span) {
+        // If we're edition 2024 or within a TAIT or RPITIT, *and* there is no
+        // `use<>` statement to override the default capture behavior, then
+        // capture all of the in-scope lifetimes.
+        if (self.always_capture_in_scope || span.at_least_rust_2024())
+            && !bounds.iter().any(|bound| match bound {
+                GenericBound::Use(..) => true,
+                _ => false,
+            })
+        {
+            for (ident, id, _) in self.resolver.extra_lifetime_params(opaque_ty_node_id) {
```

Comment:
Right so IIUC, we can no longer `take` the extra lifetime params because if we did, they would no longer be available when lowering any nested opaques (here: `impl Sized`) after having lowered any containing opaques (here: `impl IntoIterator<Item = impl Sized>`).

Reply:
Yep, we need to be able to access these extra lifetimes both when walking the bounds of the parent opaque, and also lowering the child opaque itself. I have no idea why we were using `take` before, it's not like we really enforced it.